src/reactjs/QueryClientProvider.tsx

```diff
@@ -41,24 +49,41 @@ export const useQueryClient = () => {
   return queryClient
 }
 
-export interface QueryClientProviderProps {
+type QueryClientProviderPropsBase = {
   client: QueryClient
+}
+type QueryClientProviderPropsWithContext = ContextOptions &
+  QueryClientProviderPropsBase
+type QueryClientProviderPropsWithContextSharing = {
   contextSharing?: boolean
+} & QueryClientProviderPropsBase
+
+export type QueryClientProviderProps =
+  | QueryClientProviderPropsWithContext
+  | QueryClientProviderPropsWithContextSharing
+
+const propsHasContext = (
+  props: QueryClientProviderProps
+): props is QueryClientProviderPropsWithContext => {
+  return (props as QueryClientProviderPropsWithContext).context !== undefined
 }
 
-export const QueryClientProvider: React.FC<QueryClientProviderProps> = ({
-  client,
-  contextSharing = false,
-  children,
-}) => {
+export const QueryClientProvider: React.FC<QueryClientProviderProps> = props => {
+  const { client, children } = props
   React.useEffect(() => {
     client.mount()
     return () => {
       client.unmount()
     }
   }, [client])
 
-  const Context = getQueryClientContext(contextSharing)
+  const contextSharing = propsHasContext(props)
+    ? false
+    : props.contextSharing || false
+
+  const Context = propsHasContext(props)
+    ? getQueryClientContext(props.context, false)
+    : getQueryClientContext(undefined, contextSharing)
```

Comment:
is that really necessary? I think we could:
- define `context?: never` on `QueryClientProviderPropsWithContextSharing`
- just pass all props to `getQueryClientContext` an destruct there
- then, the early return for `if (context)` should work just the same because if we have a context, we can't have contextSharing.

Reply:
Yep, absolutely agree. Great callout on `context?: never`. The existing types were not enforcing the mutual exclusivity.  

Changes have been made and pushed up.

Reply:
Done